api/backend/ecs/resource_manager.go

```diff
@@ -0,0 +1,201 @@
+package ecsbackend
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/quintilesims/layer0/api/backend/ecs/id"
+	"github.com/quintilesims/layer0/api/scheduler/resource"
+	"github.com/quintilesims/layer0/common/aws/autoscaling"
+	"github.com/quintilesims/layer0/common/aws/ec2"
+	"github.com/quintilesims/layer0/common/aws/ecs"
+	"github.com/quintilesims/layer0/common/logutils"
+	"github.com/zpatrick/go-bytesize"
+	"strconv"
+)
+
+type ECSResourceManager struct {
+	ECS         ecs.Provider
+	Autoscaling autoscaling.Provider
+	logger      *logrus.Logger
+}
+
+func NewECSResourceManager(e ecs.Provider, a autoscaling.Provider) *ECSResourceManager {
+	return &ECSResourceManager{
+		ECS:         e,
+		Autoscaling: a,
+		logger:      logutils.NewStandardLogger("ECS Resource Manager"),
+	}
+}
+
+func (r *ECSResourceManager) GetProviders(environmentID string) ([]*resource.ResourceProvider, error) {
+	ecsEnvironmentID := id.L0EnvironmentID(environmentID).ECSEnvironmentID()
+
+	instanceARNs, err := r.ECS.ListContainerInstances(ecsEnvironmentID.String())
+	if err != nil {
+		return nil, err
+	}
+
+	resourceProviders := []*resource.ResourceProvider{}
+	if len(instanceARNs) > 0 {
+		instances, err := r.ECS.DescribeContainerInstances(ecsEnvironmentID.String(), instanceARNs)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, instance := range instances {
+			provider, ok := r.getResourceProvider(ecsEnvironmentID, instance)
+			if ok {
+				resourceProviders = append(resourceProviders, provider)
+			}
+		}
+	}
+
+	return resourceProviders, nil
+}
+
+func (r *ECSResourceManager) getResourceProvider(ecsEnvironmentID id.ECSEnvironmentID, instance *ecs.ContainerInstance) (*resource.ResourceProvider, bool) {
+	instanceID := pstring(instance.Ec2InstanceId)
+	if pstring(instance.Status) != "ACTIVE" {
+		r.logger.Warnf("Instance %d is not in the 'ACTIVE' state", instanceID)
+		return nil, false
+	}
+
+	// todo: what if instance.AgentConnected == false?
```

Comment:
Let's log that the provider is not available and let it sit (in the case of Splunk or other long-running services)